addons/sms/models/sms_sms.py

```diff
@@ -96,20 +97,25 @@ def action_set_error(self, failure_type):
     def action_set_outgoing(self):
         self._update_sms_state_and_trackers('outgoing', failure_type=False)
 
-    def send(self, unlink_failed=False, unlink_sent=True, auto_commit=False, raise_exception=False):
+    def send(self, unlink_failed=False, unlink_sent=True, auto_commit=None, raise_exception=False):
         """ Main API method to send SMS.
 
+        This contacts an external server. If the transaction fails, it may be
+        retried which can result in sending multiple SMS messages!
+
           :param unlink_failed: unlink failed SMS after IAP feedback;
           :param unlink_sent: unlink sent SMS after IAP feedback;
-          :param auto_commit: commit after each batch of SMS;
+          :param auto_commit: deprecated, use CRON if needed to commit
           :param raise_exception: raise if there is an issue contacting IAP;
         """
-        self = self.filtered(lambda sms: sms.state == 'outgoing' and not sms.to_delete)
-        for batch_ids in self._split_batch():
-            self.browse(batch_ids)._send(unlink_failed=unlink_failed, unlink_sent=unlink_sent, raise_exception=raise_exception)
-            # auto-commit if asked except in testing mode
-            if auto_commit is True and not modules.module.current_test:
-                self._cr.commit()
+
+        if auto_commit is not None:
+            warnings.warn("Since 19.0, parameter auto_commit is deprecated", DeprecationWarning)
+        domain = [('state', '=', 'outgoing'), ('to_delete', '!=', True)]
+        records = self.try_lock_for_update().filtered_domain(domain)
```

Comment:
Shouldn't we filter then lock ? That way we don't lock unnecessary records

Reply:
No because something else may update the record after we filtered it (even if that may be very rare).
Basically, before trying to send, we lock what we are given.

Reply:
Seems unfortunate that `try_lock_for_update` already takes in a limit and returns a new recordset, but doesn't allow using a domain for filtering. Looks like it would be a fairly trivial and useful change, to my naive eyes at least :shrug:

Reply:
Good feedback, but I don't think we will make that change quite soon. It is not trivial, but not extremely hard either (thinks to lookout for https://stackoverflow.com/questions/6998372/sql-semantics-for-for-update-join). Most of the time, this method is used without a limit.

Reply:
Ah indeed hadn't thought of relational domains. Well, too bad. Indeed probably not common enough to warrant the effort ^^'